### PR TITLE
Add text classification model

### PR DESCRIPTION
**Pull request by kernelmachine**
This PR adds a generic text classifier to AllenNLP.

After embedding the text into a text field, this model will optionally encode the embeddings with a Seq2VecEncoder and then pass the embeddings to a linear classification layer, which projects into the label space.

### CODE REVIEW

File: allennlp/models/classifier.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        else:
+            self._encoder = None
+            self._classifier_input_dim = self._text_field_embedder.get_output_dim()
+
+        if dropout:
+            self._dropout = torch.nn.Dropout(dropout)
+        else:
+            self._dropout = None
+
+        self._num_labels = vocab.get_vocab_size(namespace="labels")
```
Comment (joelgrus) [inline allennlp/models/classifier.py]: I think this should be more configurable, definitely you should be able to specify a different namespace, possibly you should be allowed to set it explicitly as well?

File: allennlp/models/classifier.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        # pylint: disable=arguments-differ
+        """
+        Parameters
+        ----------
+        tokens : Dict[str, torch.LongTensor]
+            From a ``TextField``
+        label : torch.IntTensor, optional (default = None)
+            From a ``LabelField``
+        metadata : ``List[Dict[str, Any]]``, optional, (default = None)
+            Metadata to persist
```
Comment (joelgrus) [inline allennlp/models/classifier.py]: but you're not actually persisting the metadata?

File: allennlp/models/classifier.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            probabilities of the label.
+        loss : torch.FloatTensor, optional
+            A scalar loss to be optimised.
+        """
+        embedded_text = self._text_field_embedder(tokens)
+        mask = get_text_field_mask(tokens).float()
+
+        if self._encoder:
+            embedded_text = self._encoder(embedded_text,
+                                          mask=mask)
```
Comment (joelgrus) [inline allennlp/models/classifier.py]: this could fit on one line

**Comment by kernelmachine** [comment]
@joelgrus ready for another look.

**Comment by kernelmachine** [comment]
@joelgrus updated, ready for your comments

File: allennlp/models/basic_classifier.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        loss : torch.FloatTensor, optional
+            A scalar loss to be optimised.
+        """
+        embedded_text = self._text_field_embedder(tokens)
+        mask = get_text_field_mask(tokens).float()
+
+        if self._seq2seq_encoder:
+            embedded_text = self._seq2seq_encoder(embedded_text, mask=mask)
+
+        if self._seq2vec_encoder:
```
Comment (joelgrus) [inline allennlp/models/basic_classifier.py]: isn't this required now?
